url86119f019a_solzy_hive_parse_tgz-pJ8-org_apache_hadoop_hive_ql_parse_MainJ8/src/org/apache/hadoop/hive/ql/parse/HiveParser_FromClauseParser.java

```diff
@@ -0,0 +1,6078 @@
+							list_expr.add(expr.getTree());
+							}
+							break;
+
+						default :
+							break loop21;
+						}
+					}
+
+					}
+					break;
+
+			}
+
+			RPAREN76=(Token)match(input,RPAREN,FOLLOW_RPAREN_in_tableBucketSample924); if (state.failed) return retval; 
+			if ( state.backtracking==0 ) stream_RPAREN.add(RPAREN76);
+
+			// AST REWRITE
+			// elements: denominator, expr, numerator
+			// token labels: numerator, denominator
+			// rule labels: retval
+			// token list labels: 
+			// rule list labels: expr
+			// wildcard labels: 
+			if ( state.backtracking==0 ) {
+			retval.tree = root_0;
+			RewriteRuleTokenStream stream_numerator=new RewriteRuleTokenStream(adaptor,"token numerator",numerator);
+			RewriteRuleTokenStream stream_denominator=new RewriteRuleTokenStream(adaptor,"token denominator",denominator);
+			RewriteRuleSubtreeStream stream_retval=new RewriteRuleSubtreeStream(adaptor,"rule retval",retval!=null?retval.getTree():null);
+			RewriteRuleSubtreeStream stream_expr=new RewriteRuleSubtreeStream(adaptor,"token expr",list_expr);
+			root_0 = (CommonTree)adaptor.nil();
+			// 153:149: -> ^( TOK_TABLEBUCKETSAMPLE $numerator $denominator ( $expr)* )
+			{
+				// FromClauseParser.g:153:152: ^( TOK_TABLEBUCKETSAMPLE $numerator $denominator ( $expr)* )
+				{
+				CommonTree root_1 = (CommonTree)adaptor.nil();
+				root_1 = (CommonTree)adaptor.becomeRoot((CommonTree)adaptor.create(TOK_TABLEBUCKETSAMPLE, "TOK_TABLEBUCKETSAMPLE"), root_1);
+				adaptor.addChild(root_1, stream_numerator.nextNode());
+				adaptor.addChild(root_1, stream_denominator.nextNode());
+				// FromClauseParser.g:153:201: ( $expr)*
+				while ( stream_expr.hasNext() ) {
+					adaptor.addChild(root_1, stream_expr.nextTree());
+				}
+				stream_expr.reset();
+
+				adaptor.addChild(root_0, root_1);
+				}
+
+			}
+
+
+			retval.tree = root_0;
+			}
+
+			}
+
+			retval.stop = input.LT(-1);
+
+			if ( state.backtracking==0 ) {
+			retval.tree = (CommonTree)adaptor.rulePostProcessing(root_0);
+			adaptor.setTokenBoundaries(retval.tree, retval.start, retval.stop);
+			}
+			if ( state.backtracking==0 ) { gParent.msgs.pop(); }
+		}
+
+		catch (RecognitionException e) {
+		  throw e;
+		}
+
+		finally {
+			// do for sure before leaving
+		}
+		return retval;
+	}
+	// $ANTLR end "tableBucketSample"
+
+
+	public static class splitSample_return extends ParserRuleReturnScope {
+		CommonTree tree;
+		@Override
+		public CommonTree getTree() { return tree; }
+	};
+
+
+	// $ANTLR start "splitSample"
+	// FromClauseParser.g:156:1: splitSample : ( KW_TABLESAMPLE LPAREN (numerator= Number ) (percent= KW_PERCENT | KW_ROWS ) RPAREN -> {percent != null}? ^( TOK_TABLESPLITSAMPLE TOK_PERCENT $numerator) -> ^( TOK_TABLESPLITSAMPLE TOK_ROWCOUNT $numerator) | KW_TABLESAMPLE LPAREN (numerator= ByteLengthLiteral ) RPAREN -> ^( TOK_TABLESPLITSAMPLE TOK_LENGTH $numerator) );
+	public final HiveParser_FromClauseParser.splitSample_return splitSample() throws RecognitionException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 42. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.